include/up-cpp/uri/tools/Utils.h

```diff
@@ -0,0 +1,205 @@
+ * under the License.
+ * 
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2024 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef UPROTOCOL_CPP_UTILS_H
+#define UPROTOCOL_CPP_UTILS_H
+
+#include <algorithm>
+#include <cctype>
+#include <optional>
+#include <string>
+#include <array>
+#include <string_view>
+#include <arpa/inet.h>
+#include <spdlog/spdlog.h>
+#include "up-core-api/uri.pb.h"
+
+namespace uprotocol::uri {
+
+    /**
+     * isBlank
+     * is a string blank or containes only white spaces 
+     *
+     * @param str the string to check
+     *
+     * @return true if empty or with only white spaces
+     */
+    [[nodiscard]] [[maybe_unused]] static auto isBlank(
+            std::string_view str) -> bool;
+
+    /**
+     * isEmpty
+     * is Entity empty
+     *
+     * @param entity 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] static auto isEmpty(
+            uprotocol::v1::UEntity const &entity) -> bool;
+
+    /**
+     * isEmpty
+     * is Resource empty
+     *
+     * @param resource 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] static auto isEmpty(
+            uprotocol::v1::UResource const &resource) -> bool;
+
+    /**
+     * isEmpty
+     * is Authority empty
+     *
+     * @param authority 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] static auto isEmpty(
+            uprotocol::v1::UAuthority const &authority) -> bool;
+
+    /**
+     * isEmpty
+     * is Uri empty
+     *
+     * @param uri 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] static auto isEmpty(
+            uprotocol::v1::UUri const &uri) -> bool;
+
+    /**
+     * @param authority 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] auto static isResolved(
+            [[maybe_unused]] uprotocol::v1::UAuthority const &authority)
+        -> bool;
+
+    /**
+     * @param entity 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] auto static isResolved(
+            uprotocol::v1::UEntity const &entity) -> bool;
+
+    /**
+     * @param resource 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] auto static isResolved(
+            uprotocol::v1::UResource const &resource) -> bool;
+
+    /**
+     * isResolved
+     * is Uri resolved (both 
+     *
+     * @param uri 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] auto static isResolved(
+            [[maybe_unused]] v1::UUri const& uri) -> bool;
+
+
+    /**
+     * @param authority 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] auto static isLocal(
+            uprotocol::v1::UAuthority const &authority) -> bool;
+
+    /**
+     * @param authority 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] auto static isRemote(
+            uprotocol::v1::UAuthority const &authority) -> bool;
+
+    /**
+     * @param s 
+     * @param o 
+     * @return 
+     */
+    [[nodiscard]] [[maybe_unused]] auto static operator==(
```

Comment:
If we have simple composite types that need standard comparison functions and std::hash, it would be best to declare the objects as derived from std:tuple<>'s, and then expose named setter/getter. This also handy for generalized ostream printing, using std::apply.

Reply:
Previously, I'd brought up hashing of objects. Given my more recent understanding that those up-client-zenoh-cpp associative data structs are not necessary, any prior comments I made about std::hash should be ignored.